analysers/analyser_osmosis_highway_deadend.py

```diff
@@ -206,6 +206,43 @@
   r.nid IS NULL
 """
 
+sql40 = """
+SELECT
+  wid,
+  nid,
+  (SELECT ST_AsText(geom) FROM nodes WHERE id = nid)
+FROM
+  (
+  SELECT
+    id AS wid,
+    linestring,
+    nodes
+  FROM
+    highways as drivethroughs
+  WHERE
+    drivethroughs.highway = 'service' AND
+    drivethroughs.tags?'service' AND
+    drivethroughs.tags->'service' = 'drive-through' AND
+    NOT drivethroughs.is_oneway
+  ) as highways_drivethrough
+  JOIN highway_ends AS way_ends ON
+    highways_drivethrough.linestring && way_ends.linestring AND
+    way_ends.nid = ANY(highways_drivethrough.nodes)
+  LEFT JOIN highways AS other_highways ON
+    way_ends.nid = ANY(other_highways.nodes) AND
+    other_highways.id != highways_drivethrough.wid
+  JOIN nodes ON
+    nodes.id = way_ends.nid AND
+    (NOT nodes.tags?'highway' OR (
+      nodes.tags->'highway' != 'turning_circle' AND
+      nodes.tags->'highway' != 'turning_loop' AND
+      nodes.tags->'highway' != 'mini_roundabout'
+    )) AND
+    (NOT nodes.tags?'entrance' OR nodes.tags->'entrance' = 'no') -- i.e. indoor part not drawn
+WHERE
+  other_highways.id IS NULL
```

Comment:
Better SQL, simplest. Also join on highway_ends ends geom (point) and not linestring.

Better explain
```diff
-Nested Loop  (cost=963.52..57287.91 rows=1 width=48)
+Nested Loop  (cost=963.81..42095.65 rows=1 width=48)
```
@Famlam please review my suggest SQL code anyway. I do not have validated the result.


```suggestion
SELECT
  drivethroughs.id,
  nid,
  ST_AsText(nodes.geom)
FROM
  highways AS drivethroughs
  JOIN highway_ends ON
    drivethroughs.linestring && highway_ends.geom AND
    highway_ends.nid = ANY(drivethroughs.nodes)
  LEFT JOIN highways AS other_highways ON
    other_highways.linestring && highway_ends.geom AND
    highway_ends.nid = ANY(other_highways.nodes) AND
    other_highways.id != drivethroughs.id
  JOIN nodes ON
    nodes.id = highway_ends.nid AND
    (NOT nodes.tags?'highway' OR (
      nodes.tags->'highway' != 'turning_circle' AND
      nodes.tags->'highway' != 'turning_loop' AND
      nodes.tags->'highway' != 'mini_roundabout'
    )) AND
    (NOT nodes.tags?'entrance' OR nodes.tags->'entrance' = 'no') -- i.e. indoor part not drawn
WHERE
  drivethroughs.highway = 'service' AND
  drivethroughs.tags?'service' AND
  drivethroughs.tags->'service' = 'drive-through' AND
  NOT drivethroughs.is_oneway AND
  other_highways.id IS NULL```

Reply:
Hi @frodrigo thank you very much!
I tested your code on my local computer and it works perfect. (I checked with the same countries as in the initial message and the results are identical to the results of my initial code)

Would you be able to edit your commit suggestion to remove the three ``` at the end? Then it can be directly committed via the suggestion

Reply:
> Would you be able to edit your commit suggestion to remove the three ``` at the end? Then it can be directly committed via the suggestion
done

Reply:
Applied, thanks!